feat: Added slack chat integration.

osala-eng opened the pull request:
## Description

Slack chat bot integration

Implements #227 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration by modifying the list below.

- [ ] Test A
- [ ] Test B
- [ ] Test C

**Test Configuration**:

Please describe the test setup. List them below as bullet points.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] The commit message follows the convention of this project

osala-eng commented:
@janaka  I've setup slack chat bot. you can review this and tell me what you think.

I've also added the sample chat bot to the Docq.AI's slack work space you can try that out aswell.

janaka commented:
Great. I'll take a look.

janaka commented:
How do I do RAG? Or you waiting for the updated Web API live?

osala-eng commented:
> How do I do RAG? Or you waiting for the updated Web API live?

This just implements chat comletion as described in #227, once we get this approved i'll use it to implement Retrival in #228

janaka commented:
great, I'll take a look at the code this weekend. 

I shipped the new Web API btw.

janaka reviewed (commented):
Looks like the whole integration is webhooks based? 

This code should be part of the API I think. It also should be designed as a general webhook handler ideally.

janaka commented on source/docq/integrations/slack.py:
```diff
@@ -0,0 +1,213 @@
+
+from .models import SlackChannel, SlackInstallation
+
+SQL_CREATE_DOCQ_SLACK_APPLICATIONS_TABLE = """
+CREATE TABLE IF NOT EXISTS docq_slack_installations (
+    id INTEGER PRIMARY KEY,
+    app_id TEXT NOT NULL,
+    team_id TEXT NOT NULL,
+    team_name TEXT NOT NULL,
+    org_id INTEGER NOT NULL,
+    space_group_id INTEGER,
+    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    FOREIGN KEY (org_id) REFERENCES orgs(id),
+    FOREIGN KEY (space_group_id) REFERENCES space_groups(id),
+    UNIQUE (app_id, team_id, org_id)
+);
+"""
```
Can you explain the data in this table pls. it's not obvious to me.

I don't love having Slack specific tables. It's OK that you started like this to get the whole thing up and running end to end. 

I'm wondering if we should iterate to a similar/same pattern as config for data sources.

osala-eng replied:
Added comments on this.

janaka commented on source/docq/setup.py:
```diff
@@ -38,6 +39,7 @@ def init() -> None:
         manage_assistants._init()
         db_migrations.run() # run db migrations after all tables are created
         services._init()
+        slack._init()
```
pls put this inside and integrations._init() and call that here.

osala-eng replied:
Done

janaka commented on source/docq/support/auth_utils.py:
```diff
@@ -45,15 +45,18 @@ def init_session_cache() -> None:
     """Set client cookie for authentication."""
-    try:
+    if name is None:
+        name = SESSION_COOKIE_NAME
```
this shouldn't be here now right? That value should be passed in.

and name should not be Optional therefore.

osala-eng replied:
Created a separate method for this.

janaka commented on source/docq/support/auth_utils.py:
```diff
@@ -45,15 +45,18 @@ def init_session_cache() -> None:
-    try:
+    if name is None:
+        name = SESSION_COOKIE_NAME
+    if expiry is None:
```
thinking it makes more sense for this function param to be`ttl` with the default  set to `TTL_HOURS`. Also renamed to COOKIE_DEFAULT_TTL_HOURS.

janaka replied:
unless I'm missing a use case.

osala-eng replied:
Done.

janaka commented on web/api/integration/slack/slack_application.py:
```diff
@@ -0,0 +1,53 @@
+    return failure_args.default.failure(failure_args)
+
+
+slack_app = App(
+    oauth_flow=SlackOAuthFlow.sqlite3(
+        database=get_sqlite_shared_system_file(),
+        install_path="/api/integration/slack/v1/install",
+        redirect_uri_path="/api/integration/slack/v1/oauth_redirect",
+        client_id=CLIENT_ID,
+        client_secret=CLIENT_SECRET,
+        scopes=SCOPES,
+        user_scopes=USER_SCOPES,
+        callback_options=CallbackOptions(success=success_callback, failure=failure_callback),
+    )
+)
```
what's going on here? this seems off with sqlite3 being called directly. Database calls should all be contained strictly within the data layer.

osala-eng replied:
Moved this to data layer.

janaka commented on web/api/integration/slack/oauth_flow.py:
```diff
@@ -0,0 +1,129 @@
+    """Custom slack oauth flow."""
+
+    @classmethod
+    def sqlite3(
```
same comment here re: sqlite

janaka commented on web/api/integration/slack/chat_handler.py:
```diff
@@ -0,0 +1,22 @@
+"""Slack chat action handler."""
```
is `chat` a category of Slack events? so the pattern is we'll create a separate hanlder file to handle each category?

janaka replied:
Generally worth referencing links to Slack docs.

janaka reviewed (changes requested):
See comments. Please respond inline. 

I've not paid attention to the integrations setting / install stuff because it seems you are still working on the changes we discussed, moving those to the Admin Section.

osala-eng commented:
@janaka I've updated the changes as requested and implemented end to end slack chat with docq bot using RAG.

![image](https://github.com/docqai/docq/assets/64925863/728dbde3-eed7-4075-88e9-06765a72fce7)

janaka commented:
Great. Will take a look tomorrow.

janaka commented:
@osala-eng sorry, I merged the Llama Index update last night. Can you rebase and check all works. Mainly with the bits that bit the Web API

osala-eng commented:
> @osala-eng sorry, I merged the Llama Index update last night. Can you rebase and check all works. Mainly with the bits that bit the Web API

Okay, let me rebase and test this.

janaka commented:
great - do you have one or two tests for any of the new functionality build
for Slack and Chrome. Would be handy to know there's something.

On Sun, 24 Mar 2024 at 16:31, Jashon Osala ***@***.***> wrote:

> @osala-eng <https://github.com/osala-eng> sorry, I merged the Llama Index
> update last night. Can you rebase and check all works. Mainly with the bits
> that bit the Web API
>
> Okay, let me rebase and test this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docqai/docq/pull/236#issuecomment-2016861975>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAMUQYUMGEMXMN3SJWLRRTYZ35WVAVCNFSM6AAAAABENPVKL6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMJWHA3DCOJXGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

osala-eng commented:
@janaka There is a problem with [main](https://github.com/docqai/docq/tree/main) - I'll rebase this branch once that is sorted.

janaka commented:
What's the issue on main branch? Seem to be ok, test passed etc.

janaka commented:
Was that to do with your local virtual environment?

osala-eng commented:
Yes, it was my local env.

On Wed, 27 Mar 2024, 23:02 Janaka Abeywardhana, ***@***.***>
wrote:

> Was that to do with your local virtual environment?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docqai/docq/pull/236#issuecomment-2023862456>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APPLBJ5A64WIP7L3OPUJVATY2MJVVAVCNFSM6AAAAABENPVKL6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRTHA3DENBVGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

osala-eng commented:
This is ready for you now.

On Fri, Mar 29, 2024 at 5:52 PM Janaka Abeywardhana <
***@***.***> wrote:

> @osala-eng <https://github.com/osala-eng> what's the state of this PR?
> are you still making changes or over to me?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/docqai/docq/pull/236#issuecomment-2027346078>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APPLBJ5FVRPI5WCVXX6DJITY2VWZXAVCNFSM6AAAAABENPVKL6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMRXGM2DMMBXHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

janaka reviewed (commented):
@osala-eng I pushed a bunch of refactors. Any chance you can do a little quick test.

osala-eng commented:
> @osala-eng I pushed a bunch of refactors. Any chance you can do a little quick test.

Okay let me do a quick test and get back to you.

osala-eng commented on source/docq/integrations/slack/slack_oauth_flow.py:
```diff
@@ -0,0 +1,128 @@
+                state_expiration_seconds=state_expiration_seconds,
+            ),
+        )
+
+    # def get_cookie(self: Self, name: str, cookies: Optional[str | Sequence[str]]) -> Optional[str]:
+    #     """Get a cookie."""
+    #     from docq.support.auth_utils import decrypt_cookie_value
+
+    #     if not cookies:
+    #         return None
+
+    #     if isinstance(cookies, str):
+    #         cookies = [cookies]
+    #     for cookie in cookies:
+    #         for item in cookie.split(";"):
+    #             key, value = item.split("=", 1)
+    #             if key.strip() == name:
+    #                 return decrypt_cookie_value(value)
+    #     return None
+
+    # def save_docq_slack_installation(self: Self, request: BoltRequest, installation: Installation) -> None:
+    #     """Save a Docq slack installation."""
+    #     docq_slack_app_state =  self.get_cookie("docq_slack_app_state", request.headers.get("cookie"))
+    #     if docq_slack_app_state is not None:
+    #         _, selected_org_id = docq_slack_app_state.split(":")
+    #         create_docq_slack_installation(installation, int(selected_org_id))
+    #     else:
+    #         raise BoltError("Login to Docq before installing the slack app.")
```
@janaka I see you've commented out some of the code.

janaka replied:
Seemed like those weren't being called. Maybe my VS Code search wasn't working properly. If you uncomment that does it work.

janaka replied:
great! mind adding a few comments to document that logic.

Assume all is working as expected at this point?

osala-eng replied:
Yes, I've tested and all is working as expected.

osala-eng replied:
Let me add a few comments on this.

osala-eng reviewed (commented):
This dosent work fully as some of the functionality has been disabled.

osala-eng commented:
@janaka This is now ready.

janaka commented:
@osala-eng did you write down anything / ref a link on how the values for the following were generated?


SLACK_CLIENT_ID
SLACK_CLIENT_SECRET
SLACK_SIGNING_SECRET

janaka commented on source/docq/integrations/slack/slack_application.py:
```diff
@@ -0,0 +1,37 @@
+CLIENT_ID = os.environ.get("SLACK_CLIENT_ID")
+CLIENT_SECRET = os.environ.get("SLACK_CLIENT_SECRET")
+SCOPES = ["app_mentions:read", "im:history", "chat:write", "channels:read", "groups:read", "im:read", "mpim:read"]
+USER_SCOPES = ["admin"]
```
@osala-eng wondering if we can have a lesser scope here.

osala-eng replied:
The only scope that you can remove without affecting the functionality is `admin` but then anyone within a slack team will be able to install the application. Currently, only admins can install the application.

For the rest, we are mostly requesting read access, except for chat, where we need write access to enable the bot to respond to mentions.

janaka replied:
I've removed the user_scopes `admin` for now. As I understood from the docs user scopes allow the bot to do things on behalf of a user / as a user with those perms. maybe I've misunderstood. Will revisit.

janaka commented:
> @osala-eng did you write down anything / ref a link on how the values for the following were generated?
> 
> SLACK_CLIENT_ID SLACK_CLIENT_SECRET SLACK_SIGNING_SECRET

@osala-eng I just remembered about the other repo with the manifest file which answer my original question sufficiently.

New related question about the env var names. You went with the SDK built-in names rather than our usual pattern of prefixing with "DOCQ_". Was that intentional for reasons like it's the only way to use the SDK? I wanted to check before refactoring to save myself some pain :)

osala-eng commented:
> > @osala-eng did you write down anything / ref a link on how the values for the following were generated?
> > SLACK_CLIENT_ID SLACK_CLIENT_SECRET SLACK_SIGNING_SECRET
> 
> @osala-eng I just remembered about the other repo with the manifest file which answer my original question sufficiently.
> 
> New related question about the env var names. You went with the SDK built-in names rather than our usual pattern of prefixing with "DOCQ_". Was that intentional for reasons like it's the only way to use the SDK? I wanted to check before refactoring to save myself some pain :)

Yes, you can customize the env var names, then pass them explicitly when instanciating the `SlackOauthFlow` class. By default, slack_sdk looks for the names provided in their documentation.
